File: games/breakpoint/src/physics/Vec.ts
/**
 * Minimal vector maths for the physics core.
 *
 * Deliberately plain objects rather than classes with methods: the simulation
 * allocates nothing per step in its hot loops, and plain records serialise
 * straight into shot records without a custom encoder.
 *
 * Convention used everywhere in `physics/`: the cloth is the z = 0 plane, +z is
 * up, ball centres sit at z = R. Linear motion is therefore two-dimensional
 * (`Vec2`) while angular velocity needs all three axes (`Vec3`) — ωz is English.
 */

export interface Vec2 {
  x: number;
  y: number;
}

export interface Vec3 {
  x: number;
  y: number;
  z: number;
}

export const v2 = (x = 0, y = 0): Vec2 => ({ x, y });
export const v3 = (x = 0, y = 0, z = 0): Vec3 => ({ x, y, z });

export const clone2 = (a: Vec2): Vec2 => ({ x: a.x, y: a.y });
export const clone3 = (a: Vec3): Vec3 => ({ x: a.x, y: a.y, z: a.z });

export const add2 = (a: Vec2, b: Vec2): Vec2 => ({ x: a.x + b.x, y: a.y + b.y });
export const sub2 = (a: Vec2, b: Vec2): Vec2 => ({ x: a.x - b.x, y: a.y - b.y });
export const scale2 = (a: Vec2, s: number): Vec2 => ({ x: a.x * s, y: a.y * s });
export const dot2 = (a: Vec2, b: Vec2): number => a.x * b.x + a.y * b.y;
export const len2 = (a: Vec2): number => Math.hypot(a.x, a.y);
export const lenSq2 = (a: Vec2): number => a.x * a.x + a.y * a.y;

/** Normalise, returning the zero vector for a degenerate input rather than NaN. */
export function norm2(a: Vec2): Vec2 {
  const l = Math.hypot(a.x, a.y);
  if (l < 1e-12) return { x: 0, y: 0 };
  return { x: a.x / l, y: a.y / l };
}

/** Rotate 90° counter-clockwise — the "left of travel" direction. */
export const perp2 = (a: Vec2): Vec2 => ({ x: -a.y, y: a.x });

export const add3 = (a: Vec3, b: Vec3): Vec3 => ({ x: a.x + b.x, y: a.y + b.y, z: a.z + b.z });
export const sub3 = (a: Vec3, b: Vec3): Vec3 => ({ x: a.x - b.x, y: a.y - b.y, z: a.z - b.z });
export const scale3 = (a: Vec3, s: number): Vec3 => ({ x: a.x * s, y: a.y * s, z: a.z * s });
export const dot3 = (a: Vec3, b: Vec3): number => a.x * b.x + a.y * b.y + a.z * b.z;
export const len3 = (a: Vec3): number => Math.hypot(a.x, a.y, a.z);

export const cross3 = (a: Vec3, b: Vec3): Vec3 => ({
  x: a.y * b.z - a.z * b.y,
  y: a.z * b.x - a.x * b.z,
  z: a.x * b.y - a.y * b.x,
});

export function norm3(a: Vec3): Vec3 {
  const l = Math.hypot(a.x, a.y, a.z);
  if (l < 1e-12) return { x: 0, y: 0, z: 0 };
  return { x: a.x / l, y: a.y / l, z: a.z / l };
}

export const clamp = (v: number, lo: number, hi: number): number => (v < lo ? lo : v > hi ? hi : v);

/** True only for real, representable numbers — the NaN/Infinity tripwire. */
export const isFiniteNum = (v: number): boolean => Number.isFinite(v);

File: games/breakpoint-unity/Assets/BREAKPOINT/Runtime/Simulation/Vec.cs
using System;

namespace Breakpoint.Simulation
{
    /// <summary>
    /// Minimal vector maths for the simulation core.
    ///
    /// Everything here is <c>double</c>, deliberately. The TypeScript reference
    /// runs on JavaScript Numbers, which are IEEE-754 binary64; using
    /// <c>float</c> inside the authoritative simulation would silently change
    /// results and make parity with the reference unachievable. Unity is welcome
    /// to take single-precision copies for rendering — see the presentation
    /// bridge — but it never hands them back.
    ///
    /// These are readonly structs so they live on the stack and cost no
    /// allocation inside the 120 Hz loop.
    /// </summary>
    public readonly struct Vec2 : IEquatable<Vec2>
    {
        public readonly double X;
        public readonly double Y;

        public Vec2(double x, double y)
        {
            X = x;
            Y = y;
        }

        public static readonly Vec2 Zero = new Vec2(0.0, 0.0);

        public double Length => Math.Sqrt(X * X + Y * Y);
        public double LengthSquared => X * X + Y * Y;
        public bool IsFinite => Numeric.IsFinite(X) && Numeric.IsFinite(Y);

        public static Vec2 operator +(Vec2 a, Vec2 b) => new Vec2(a.X + b.X, a.Y + b.Y);
        public static Vec2 operator -(Vec2 a, Vec2 b) => new Vec2(a.X - b.X, a.Y - b.Y);
        public static Vec2 operator *(Vec2 a, double s) => new Vec2(a.X * s, a.Y * s);
        public static Vec2 operator *(double s, Vec2 a) => new Vec2(a.X * s, a.Y * s);
        public static Vec2 operator -(Vec2 a) => new Vec2(-a.X, -a.Y);

        public static double Dot(Vec2 a, Vec2 b) => a.X * b.X + a.Y * b.Y;

        /// <summary>Normalise, returning zero for a degenerate input rather than NaN.</summary>
        public Vec2 Normalized()
        {
            double length = Math.Sqrt(X * X + Y * Y);
            if (length < 1e-12) return Zero;
            return new Vec2(X / length, Y / length);
        }

        public bool Equals(Vec2 other) => X.Equals(other.X) && Y.Equals(other.Y);
        public override bool Equals(object obj) => obj is Vec2 other && Equals(other);
        public override int GetHashCode() => unchecked((X.GetHashCode() * 397) ^ Y.GetHashCode());
        public override string ToString() => "(" + X.ToString("R") + ", " + Y.ToString("R") + ")";
    }

    /// <summary>
    /// Three-component vector, used for angular velocity.
    ///
    /// Linear motion is planar (<see cref="Vec2"/>) because balls never leave
    /// the cloth, but spin needs all three axes: X and Y carry topspin and
    /// backspin, Z carries English.
    /// </summary>
    public readonly struct Vec3 : IEquatable<Vec3>
    {
        public readonly double X;
        public readonly double Y;
        public readonly double Z;

        public Vec3(double x, double y, double z)
        {
            X = x;
            Y = y;
            Z = z;
        }

        public static readonly Vec3 Zero = new Vec3(0.0, 0.0, 0.0);

        public double Length => Math.Sqrt(X * X + Y * Y + Z * Z);
        public double LengthSquared => X * X + Y * Y + Z * Z;
        public bool IsFinite => Numeric.IsFinite(X) && Numeric.IsFinite(Y) && Numeric.IsFinite(Z);

        public static Vec3 operator +(Vec3 a, Vec3 b) => new Vec3(a.X + b.X, a.Y + b.Y, a.Z + b.Z);
        public static Vec3 operator -(Vec3 a, Vec3 b) => new Vec3(a.X - b.X, a.Y - b.Y, a.Z - b.Z);
        public static Vec3 operator *(Vec3 a, double s) => new Vec3(a.X * s, a.Y * s, a.Z * s);
        public static Vec3 operator *(double s, Vec3 a) => new Vec3(a.X * s, a.Y * s, a.Z * s);

        public static double Dot(Vec3 a, Vec3 b) => a.X * b.X + a.Y * b.Y + a.Z * b.Z;

        public static Vec3 Cross(Vec3 a, Vec3 b) => new Vec3(
            a.Y * b.Z - a.Z * b.Y,
            a.Z * b.X - a.X * b.Z,
            a.X * b.Y - a.Y * b.X);

        public bool Equals(Vec3 other) => X.Equals(other.X) && Y.Equals(other.Y) && Z.Equals(other.Z);
        public override bool Equals(object obj) => obj is Vec3 other && Equals(other);

        public override int GetHashCode()
        {
            unchecked
            {
                int hash = X.GetHashCode();
                hash = (hash * 397) ^ Y.GetHashCode();
                return (hash * 397) ^ Z.GetHashCode();
            }
        }

        public override string ToString() =>
            "(" + X.ToString("R") + ", " + Y.ToString("R") + ", " + Z.ToString("R") + ")";
    }

    /// <summary>Numeric helpers shared across the simulation.</summary>
    public static class Numeric
    {
        /// <summary>
        /// True only for a real, representable number.
        ///
        /// This is the NaN/Infinity tripwire. It exists as one function so the
        /// check cannot drift between the places that guard state.
        /// </summary>
        public static bool IsFinite(double value) => !double.IsNaN(value) && !double.IsInfinity(value);

        public static double Clamp(double value, double min, double max)
        {
            if (value < min) return min;
            if (value > max) return max;
            return value;
        }
    }
}
